File: problems/0202.IsHappy/IsHappy.java
import java.util.HashSet;
import java.util.Set;

class IsHappy {
    public boolean isHappy(int n) {
        Set<Integer> seen = new HashSet<>();
        
        while (n != 1 && !seen.contains(n)) {
            seen.add(n);
            n = getNext(n);
        }
        
        return n == 1;
    }
    
    private int getNext(int num) {
        int total = 0;
        while (num > 0) {
            int digit = num % 10;
            total += digit * digit;
            num /= 10;
        }
        return total;
    }

    public static void main(String[] args) {
        IsHappy ih = new IsHappy();
        System.out.println(ih.isHappy(19));
    }
}

File: problems/0202.IsHappy/IsHappy.go
package main

import "fmt"

func isHappy(n int) bool {
	seen := make(map[int]bool)

	for n != 1 && !seen[n] {
		seen[n] = true
		n = getNext(n)
	}

	return n == 1
}

func getNext(num int) int {
	total := 0
	for num > 0 {
		digit := num % 10
		total += digit * digit
		num /= 10
	}
	return total
}

func main() {
	fmt.Println(isHappy(19))
	fmt.Println(isHappy(2))
}
